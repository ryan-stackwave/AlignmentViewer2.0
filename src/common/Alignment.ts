import { generateUUIDv4 } from "./Utils";
import { defineNewAlignmentMode } from "./AceAlignmentMode";
import { Nucleotide } from "./Residues";
import { SequenceSorter } from "./AlignmentSorter";
import {
  AlignmentTypes,
  AminoAcidAlignmentStyle,
  NucleotideAlignmentStyle,
} from "./MolecularStyles";

export interface ISequence {
  id: string;
  sequence: string;
}

/**
 * Alignment
 * This class represents a multiple sequence alignment. During initialization,
 * positional statistics are generated about the alignment and a consensus
 * sequence is determined.
 *
 * The class should generally be instantiated from subclasses (e.g., FastaAlignemtn)
 * using static methods, e.g.,
 *     FastaAlignment.fromFileContents
 *     StockholmAlignment.fromFileContents
 */
export class Alignment {
  private uuid: string;
  private aceEditorMode: string | undefined;
  private name: string;
  private predictedNT: boolean;
  private sequences: Map<SequenceSorter, ISequence[]>;
  private querySequence: ISequence;
  private positionalLetterCounts = new Map<
    number,
    { [letter: string]: number }
  >();
  private globalAlphaLetterCounts: { [letter: string]: number } = {};
  private allUpperAlphaLettersInAlignmentSorted: string[];
  private consensus: {
    sequence: ISequence;
    statistics: {
      letter: string;
      position: number;
      occurrences: number;
    }[];
  };

  /**
   * Normalize all values in an object that contains all counts such that
   * the sum of those counts equals 1.
   * @param letterCounts an object that contains the same keys, but whose
   *                     values now sum to 1.
   * @param validLetters if provided, returned object will only contain letters
   *                     that are in this array.
   */
  private static normalizeLetterCounts(
    letterCounts: { [letter: string]: number },
    validLetters?: string[]
  ) {
    const totalCounts = Object.keys(letterCounts).reduce((acc, letter) => {
      if (!validLetters || validLetters.includes(letter)) {
        acc += letterCounts[letter];
      }
      return acc;
    }, 0);

    return Object.keys(letterCounts).reduce((acc, letter) => {
      if (!validLetters || validLetters.includes(letter)) {
        acc[letter] = letterCounts[letter] / totalCounts;
      }
      return acc;
    }, {} as { [letter: string]: number });
  }

  /**
   * Create a new Alignment object. The default query sequence will
   * be set to the first sequence in sequences. This can be changed
   * by calling setQuerySequence() after creation
   * @param sequences
   */
  public constructor(name: string, sequencesAsInput: ISequence[]) {
    this.uuid = generateUUIDv4();
    this.name = name;
    this.sequences = new Map<SequenceSorter, ISequence[]>();
    this.sequences.set(SequenceSorter.INPUT, sequencesAsInput);
    this.querySequence = sequencesAsInput[0];
    this.predictedNT = true;

    //
    //generate statistics
    //
    const start = new Date();
    const allLettersInAlignment: { [key: string]: number } = {}; //all letters in the alignment

    // aggregate stats for each position and globally including
    // the number of time each amino acid occurs
    // ** this loop takes the bulk of the initialization time
    for (
      let sequenceIdx = 0;
      sequenceIdx < sequencesAsInput.length;
      sequenceIdx++
    ) {
      const seq = sequencesAsInput[sequenceIdx].sequence;
      for (let positionIdx = 0; positionIdx < seq.length; positionIdx++) {
        const letter = seq[positionIdx];
        const letterIsAlpha = letter.match(/[a-z]/i) ? true : false;
        if (
          letterIsAlpha &&
          this.predictedNT &&
          Nucleotide.fromSingleLetterCode(letter) === Nucleotide.UNKNOWN
        ) {
          this.predictedNT = false;
        }

        const position = positionIdx; // zero based
        allLettersInAlignment[letter] = 1;

        let letterCounts = this.positionalLetterCounts.get(position);
        if (!letterCounts) {
          letterCounts = {};
          this.positionalLetterCounts.set(position, letterCounts);
        }
        if (letter in letterCounts === false) {
          letterCounts[letter] = 0;
        }
        letterCounts[letter] += 1;

        if (letterIsAlpha) {
          if (letter in this.globalAlphaLetterCounts === false) {
            this.globalAlphaLetterCounts[letter] = 0;
          }
          this.globalAlphaLetterCounts[letter] += 1;
        }
      }
    }

    //get a list of all the upper letters found in the alignment
    this.allUpperAlphaLettersInAlignmentSorted = Object.keys(
      allLettersInAlignment
    )
      .sort()
      .reduce((arr: string[], value: string) => {
        if (value.match(/[A-Z]/)) {
          // only keep letters
          arr.push(value);
        }
        return arr;
      }, []);

    //determine consensus sequence
    const consensusStats = Array.from(this.positionalLetterCounts).map(
      ([position, letterCounts]) => {
        const topLetter = Object.keys(letterCounts).reduce((a, b) => {
          const aIsLetter = a.match(/[a-z]/i);
          const bIsLetter = b.match(/[a-z]/i);
          if (aIsLetter && !bIsLetter) {
            return a;
          }
          if (bIsLetter && !aIsLetter) {
            return b;
          }
          return letterCounts[a] > letterCounts[b] ? a : b;
        });
        return {
          position: position,
          letter: topLetter,
          occurrences: letterCounts[topLetter],
        };
      }
    );
    this.consensus = {
      sequence: {
        id: "consensus",
        sequence: consensusStats.map((s) => s.letter).join(""),
      },
      statistics: consensusStats,
    };

    console.log(
      "done parsing alignment. took " +
        (new Date().getTime() - start.getTime()) +
        "ms"
    );
  }

  /**
   * Is this alignment predicted to be nucleotides?
   * @returns true if it is predicted to be nucleotide sequences.
   */
  getPredictedType(): AlignmentTypes {
    return this.predictedNT
      ? AlignmentTypes.NUCLEOTIDE
      : AlignmentTypes.AMINOACID;
  }

  /**
   * Returns a default nucleotide or amino acid style for this alignment,
   * predicted from the sequence itself.
   */
  getDefaultStyle(): AminoAcidAlignmentStyle | NucleotideAlignmentStyle {
    return this.predictedNT
      ? new NucleotideAlignmentStyle()
      : new AminoAcidAlignmentStyle();
  }

  /**
   * Get a unique uuid for this alignment.
   * @returns a uuid string for this alignment
   */
  getUUID() {
    return this.uuid;
  }

  /**
   * Get the name of this alignment
   * @returns a name for the alignment - usually the filename
   */
  getName() {
    return this.name;
  }

  /**
   * Get an ace editor mode for this alignemnt. If one doesn't exist
   * then it will be defined.
   * @returns the ace editor mode for this alignment
   */
  getAceEditorMode() {
    if (!this.aceEditorMode) {
      this.aceEditorMode = "ace/mode/" + this.uuid;
      defineNewAlignmentMode(this.aceEditorMode, this);
    }
    return this.aceEditorMode;
  }

  /**
   * Get the consensus sequence
   * @returns a position ordered array of objects that contain each letter, its number of
   *          occurrences, and the actual position number
   */
  getConsensus() {
    return this.consensus;
  }

  /**
   * Get the number of gaps at each position
   * @returns an array of gap counts, ordered by position
   */
  getGapsPerColumn(): number[] {
    return Array.from(this.positionalLetterCounts).map(
      ([position, letterCounts]) => {
        return !isNaN(letterCounts["-"]) ? letterCounts["-"] : 0;
      }
    );
  }

  /**
   * Get the number of gaps at a specific position
   * @returns the gap counts at the specified position
   */
  getGapCountAtColumn(position: number): number {
    const letterCounts = this.positionalLetterCounts.get(position);
    if (!letterCounts || !letterCounts["-"]) {
      return 0;
    }
    return letterCounts["-"];
  }

  /**
   * Get the total number of times each letter appears in the alignment.
   * @param normalize if true, will normalize the returned array such that
   *                  it sums to 1.
   * @param validLetters if provided, only those letters will be included
   *                     in the resulting counts
   */
  getGlobalAlphaLetterCounts(normalize?: boolean, validLetters?: string[]) {
    if (normalize) {
      return Alignment.normalizeLetterCounts(
        this.globalAlphaLetterCounts,
        validLetters
      );
    }
    return !validLetters
      ? this.globalAlphaLetterCounts
      : Object.keys(this.globalAlphaLetterCounts).reduce((acc, letter) => {
          //don't normalize, but remove invalid letters
          if (validLetters.includes(letter)) {
            acc[letter] = this.globalAlphaLetterCounts[letter];
          }
          return acc;
        }, {} as { [letter: string]: number });
  }

  /**
   * Get the length of the longest sequence
   * @returns the length of the largest sequence in this alignment
   */
  getMaxSequenceLength(): number {
    return this.getSequences().reduce((acc, seq) => {
      if (seq.sequence.length > acc) {
        acc = seq.sequence.length;
      }
      return acc;
    }, -1);
  }

  /**
   * Get the frequencies of all letters in this alignment for each position.
   * @param normalize if true, will normalize the returned letter counts for each
   *                  position such that each sums to 1.
   * @param validLetters if provided, only those letters will be included
   *                     in the resulting counts
   * @returns a Map whose keys are position (ordered) and values are a dictionary
   *          with [key = letter (e.g., amino acid code)] and [value = # occurrences].
   * TODO: Test parameter use cases.
   */
  getPositionalLetterCounts(normalize?: boolean, validLetters?: string[]) {
    if (normalize || validLetters) {
      return Array.from(this.positionalLetterCounts).reduce(
        (acc, [position, letterCounts]) => {
          acc.set(
            position,
            normalize
              ? Alignment.normalizeLetterCounts(letterCounts, validLetters) // normalize + remove invalid letters
              : Object.keys(letterCounts).reduce((acc2, letter) => {
                  //don't normalize, but remove invalid letters
                  if (validLetters!.includes(letter)) {
                    acc2[letter] = letterCounts[letter];
                  }
                  return acc2;
                }, {} as { [letter: string]: number })
          );
          return acc;
        },
        new Map<Number, { [letter: string]: number }>()
      );
    }
    return this.positionalLetterCounts;
  }

  /**
   * Get all the sequences in this alignment - ordered as they were input
   * @param sortBy specifies the order of the returned sequences. If it isn't
   *               provided, returns the sequences as they were input to create
   *               the Alignment object, e.g., usually fasta order.
   * @param forceReSort if provided, sequences are forced to be re-sorted (rather
   *                    than possibly cached)
   * @returns all sequences in this alignment
   */
  getSequences(sortBy?: SequenceSorter, forceReSort?: boolean): ISequence[] {
    sortBy = sortBy ? sortBy : SequenceSorter.INPUT;
    forceReSort = forceReSort ? forceReSort : false;
    if (
      sortBy !== SequenceSorter.INPUT &&
      (!this.sequences.has(sortBy) || forceReSort)
    ) {
      //cache not yet populated for this sort option
      this.sequences.set(
        sortBy,
        sortBy.sortFn(this.sequences.get(SequenceSorter.INPUT)!, this)
      );
    }
    return this.sequences.get(sortBy)!;
  }

  /**
   * Get a sorted list of all of the upper case letters in the alignment.
   * @returns a list of all letters in the alignment sorted alphabetically.
   */
  getAllUpperAlphaLettersInAlignmentSorted(): string[] {
    return this.allUpperAlphaLettersInAlignmentSorted;
  }

  /**
   * Get all the first sequence in the alignment, aka, target / query / first
   * @returns all sequences in this alignment
   */
  getQuerySequence(): ISequence {
    return this.querySequence;
  }

  /**
   * Set the query sequence
   * @param newQuery The new query sequence. Must already exist in the alignment.
   * @throws an error if the query sequence does not yet exist in the alignment
   *
   * IMPORTANT: Requires re-sorting
   */
  /*setQuerySequence(newQuery: ISequence) {
    const queryInAlignment = this.getSequences().find((seq) => {
      if (seq === newQuery) {
        return true;
      }
      return false;
    });
    if (queryInAlignment !== undefined) {
      this.querySequence = newQuery;

      //the resort could be really slow ... need to rethink.
      SequenceSorter.list.forEach((sortOption) => {
        //updated cached sorted sequences
        if (sortOption.isQuerySequenceDependent) {
          this.getSequences(sortOption, true);
        }
      });
      this.sequences.forEach(
        (sequences: ISequence[], sortBy: SequenceSorter) => {
          this.getSequences(sortBy, true); // force re-sorting, which updates cache
        }
      );
    } else {
      throw Error("query sequence does not exist in alignment");
    }
  }*/
}
