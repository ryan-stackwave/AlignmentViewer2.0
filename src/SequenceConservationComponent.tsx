import React from "react";
import Alignment from "./Alignment";

export interface ISequenceConservationComponentProps {
  alignment: Alignment;
  characterWidth: number;

  conservationPlotLoaded(node: HTMLDivElement): void;

  readonly id: string;
}

interface ISequenceConservationComponentState {
  isHovered: boolean;
}

export class SequenceConservationComponent extends React.Component<
  ISequenceConservationComponentProps,
  ISequenceConservationComponentState
> {
  private scrollerDiv?: HTMLDivElement | null;

  constructor(props: ISequenceConservationComponentProps) {
    super(props);
    this.state = {
      isHovered: false
    };
    this.handleSvgHover = this.handleSvgHover.bind(this);
  }

  handleSvgHover(isHovered: boolean) {
    this.setState({
      isHovered: isHovered
    });
  }

  /**
   * Compute the Kullback-Leibler (KL) entropy for individual columns in the alignment
   *        S = sum(pk * log(pk / qk)
   * also see scipy docs:
   *     https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.entropy.html
   *
   * @param pk Column data: an object whose keys are all letters in the column and
   *           values are normalized (sum to 1) frequency of occurrence.
   *           From scipy docs: "the (discrete) distribution. pk[i] is the
   *                              ... probability of event i"
   * @param qk an object whose keys are all letters in the alignment and values are
   *           normalized (sum to 1) frequency of occurrence.
   *           From scipy docs: "Sequence against which the relative entropy is
   *                              computed. Should be in the same format as pk."
   */
  computeEntropy(
    allLetters: string[],
    pk: { [letter: string]: number },
    qk: { [letter: string]: number },
    doLog?: boolean
  ) {
    return allLetters.reduce((acc, letter) => {
      if (letter in pk) {
        acc += pk[letter] * Math.log(pk[letter] / qk[letter]);
      }
      return acc;
    }, 0);
  }

  /**
   *
   * @param maxSeqLength
   */
  createConservationCountGs() {
    const toReturn: JSX.Element[] = [];
    const allLetters = this.props.alignment.getSortedAlphaLetters();
    const positionalLetterCounts = this.props.alignment.getPositionalLetterCounts(
      true,
      allLetters
    );
    const backgroundLetterCounts = this.props.alignment.getGlobalAlphaLetterCounts(
      true,
      allLetters
    );

    //const maximumEntropy = this.computeEntropy(allLetters, backgroundLetterCounts, backgroundLetterCounts);
    const entropies = Array.from(positionalLetterCounts).map(
      ([position, letterCounts]) => {
        return this.computeEntropy(
          allLetters,
          letterCounts,
          backgroundLetterCounts
        );
      }
    );

    //normalize between 0 and 1
    const maxEmpiricalEntropy = Math.max(...entropies);
    const normalizedEntropies = entropies.map(entropy => {
      return (entropy / maxEmpiricalEntropy) * 100;
    });
    //console.log('entropies', entropies);
    //console.log('normalizedEntropies', normalizedEntropies);

    normalizedEntropies.forEach((normalizedEntropy, idx) => {
      toReturn.push(
        <g transform={`translate(${idx},${100 - normalizedEntropy})`} key={idx}>
          <rect width={0.5} height={`${normalizedEntropy}`}></rect>
        </g>
      );
    });
    return toReturn;
  }

  /**
   * Generate a chart to depict the number of gaps in each column. The bar
   * chart is normalized to a maximum height of 100 (if the number of gaps in
   * a particular column was equal to the number of sequences in that column)
   * @param maxSeqLength The total number of possible sequences in the column.
   *                     Used to normalize the output to a maximum height of 100.
   */
  createGapCountGs(maxSeqLength: number) {
    const toReturn: JSX.Element[] = [];
    const numSequences = this.props.alignment.getSequences().length;
    const gapCounts = this.props.alignment.getGapsPerColumn();

    for (let i = 0; i < maxSeqLength; i++) {
      const gapPercentage = (gapCounts[i] / numSequences) * 100;
      toReturn.push(
        <g transform={`translate(${i + 0.5},${100 - gapPercentage})`} key={i}>
          <rect width={0.5} height={gapPercentage}></rect>
        </g>
      );
    }
    return toReturn;
  }

  divLoaded(e: HTMLDivElement | null) {
    if (e && !this.scrollerDiv) {
      //only run if first time element available
      this.scrollerDiv = e;
      this.props.conservationPlotLoaded(this.scrollerDiv);
    }
  }

  componentWillUnmount() {
    //TODO: notify parent of unloading and release dom
  }

  render() {
    if (!this.props.alignment || !this.props.characterWidth) {
      return null;
    }

    const maxSeqLength = this.props.alignment.getMaxSequenceLength();
    const totalWidth = this.props.characterWidth * maxSeqLength;
    const hoveredClass = this.state.isHovered
      ? "conservation hovered"
      : "conservation";

    return (
      <div>
        <div
          style={{
            backgroundColor: "rgba(255,255,255,0.9)",
            border: "solid 1px black",
            position: "absolute",
            padding: "3px",
            top: 0,
            right: 0,
            color: "red",
            visibility: this.state.isHovered ? "visible" : "hidden",
            zIndex: 1000
          }}
        >
          <span style={{ color: "#000000" }}>Entropy</span>
          <span style={{ color: "#b7b7b7", paddingLeft: 10 }}>&#35; Gaps</span>
        </div>

        <div id={this.props.id} ref={e => this.divLoaded(e)}>
          <div className="svg_container">
            <svg
              preserveAspectRatio="none"
              viewBox={`0 0 ${maxSeqLength} 100`}
              style={{ width: totalWidth, height: 100 }}
              className={hoveredClass}
              xmlns="http://www.w3.org/2000/svg"
              onMouseEnter={() => this.handleSvgHover(true)}
              onMouseLeave={() => this.handleSvgHover(false)}
            >
              <g className="gapCount">{this.createGapCountGs(maxSeqLength)}</g>
              <g className="entropy">{this.createConservationCountGs()}</g>
            </svg>
          </div>
        </div>
      </div>
    );
  }
}
