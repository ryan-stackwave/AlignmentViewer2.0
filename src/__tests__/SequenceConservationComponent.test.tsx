import * as React from "react";
import { shallow, default as Enzyme } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import "jest-canvas-mock";

import Alignment from "../Alignment";
import { SequenceConservationComponent } from "../SequenceConservationComponent";

describe("SequenceConservationComponent", () => {
  beforeAll(() => {
    Enzyme.configure({ adapter: new Adapter() });
  });

  it("Should render when given default props.", () => {
    const wrapper = shallow(
      <SequenceConservationComponent
        characterWidth={3}
        conservationPlotLoaded={jest.fn()}
        alignment={new Alignment("", [])}
        id={"test-sequence-logo-component"}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
