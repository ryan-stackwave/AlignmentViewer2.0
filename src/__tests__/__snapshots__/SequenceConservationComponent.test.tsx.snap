// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SequenceConservationComponent Should render when given default props. 1`] = `
<div>
  <div
    style={
      Object {
        "backgroundColor": "rgba(255,255,255,0.9)",
        "border": "solid 1px black",
        "color": "red",
        "padding": "3px",
        "position": "absolute",
        "right": 0,
        "top": 0,
        "visibility": "hidden",
        "zIndex": 1000,
      }
    }
  >
    <span
      style={
        Object {
          "color": "red",
        }
      }
    >
      Entropy
    </span>
    <span
      style={
        Object {
          "color": "green",
          "paddingLeft": 10,
        }
      }
    >
      # Gaps
    </span>
  </div>
  <div
    id="test-sequence-logo-component"
  >
    <div
      className="svg_container"
    >
      <svg
        className="conservation"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        viewBox="3 -100 -3 100"
        width={-3}
      >
        <g
          className="gapCount"
          transform="scale(1, -1)"
        />
        <g
          className="entropy"
          transform="scale(1, -1)"
        />
      </svg>
    </div>
  </div>
</div>
`;
